### ISSUE TITLE
Map zooms back to initial zoom on any state change.

### ISSUE THREAD

**Issue by myleftshoe** [googlemap-react/googlemap-react#4]
Sorry for another issue: 

Zoom the map and update some unrelated state - the map zooms and re-centers to the values specified in opts.

I think this is why tomchentw's lib has a defaultZoom and defaultCenter.

**Comment by lucifer1004** [comment]
You can avoid this issue by re-design your component tree structure. Have a look at `examples/CRA`.

When it is unavoidable, you can manually update `opts.zoom` and `opts.center` via `onZoomChanged` and `onCenterChanged`.

**Comment by myleftshoe** [comment]
@lucifer1004, thanks! doing it now. Unrelated but how do I access the map object directly?

**Comment by lucifer1004** [comment]
It is described in the [doc](https://doc.googlemap-react.com/). You can read the subsection`Direct access to Google Map objects`.
Also, you can look at the `CenterButton` component in `example/CRA`.

You can access the map object via `GoogleMapContext`:

```javascript
// Get the context state
const {state} = useContext(GoogleMapContext)

// Then do anything you want to do, but remember to check existence, because `state.map` and `state.objects` are undefined at first.
state.map && state.map.getCenter()
```

**Comment by lucifer1004** [comment]
A principle is to avoid unnecessary `MapBox` rerenders.

**Comment by lucifer1004** [comment]
Check if the new version `v0.1.1-alpha.1` helps. @myleftshoe

**Comment by myleftshoe** [comment]
@lucifer1004. Man you are fast, I'm still thinking about things and you've fixed it. It works!

**Comment by myleftshoe** [comment]
@lucifer1004. wrt Direct access to Google Map objects, I'd read that section of the docs but it wasn't clear to me as a newcomer. Your response above is much more explicit, maybe put that in the docs?

**Comment by lucifer1004** [comment]
Similar things might still happen to other components. I will memoize options for other components next.




| |
吴自华
邮箱：qqbbnease1004@126.com
|

签名由 网易邮箱大师 定制

On 03/08/2019 08:18, myleftshoe wrote:

Closed #4.

**Comment by lucifer1004** [comment]
@myleftshoe You can join the [community](https://spectrum.chat/googlemap-react) on Spectrum.

**Comment by srmaharjan** [comment]
@lucifer1004 

I am also facing same kind of issue with OverlayView component. 

You can reproduce the issue using following link

https://codesandbox.io/s/react-example-5jm3j

To reproduce the issue follow the following steps 
Once map has been loaded fully with overlay move the map or zoom in/out the map.
After that whenever we click on Counter link (to change state of count) the map re-center or zoom back to original state.

this happened only once on each reload i.e. Event though we move map after first click, map won't re-center or zoom back to original state or second or third click.

**Comment by lucifer1004** [comment]
@srmaharjan 

See this [sandbox](https://codesandbox.io/s/react-example-zipje).

The problem in your code is that <MapBox> is rerendered when the value of counter changes for the first time. I have not figured out why, but I hope that the solution above can help you out.

**Comment by lucifer1004** [comment]
The source of this issue is the `opts.mapTypeId` property from nowhere.

**Comment by lucifer1004** [comment]
Confirmed:

`new google.maps.Map(document.getElementById(mapItemId), opts)` modifies `opts`

**Comment by lucifer1004** [comment]
@srmaharjan This issue has been fixed in `v0.1.4`

**Comment by srmaharjan** [comment]
That works thank you @lucifer1004
